### ISSUE TITLE
Generic Handler implementation only handles single generic argument and breaks existing implementations 

### ISSUE THREAD

**Issue by hisuwh** [jbogard/MediatR#1047]
Support for Generic Handlers has been added in v12.3.0 in this pr: #1013

However, this is not opt-in and only supports a single generic argument.

So a setup like this is incompatible:

```csharp
public class MyRequest<T1, T2> : IRequest<T2>

public class MyRequestHandler<T1, T2> : IRequestHandler<MyRequest<T1, T2>, T2>
```

Also rather than simply not matching these requests this throws an exception on startup meaning this can't even fallback to existing code I have to handle these more complex generic handlers.

I suggest:
- if it cannot find/close a handler it should ignore it (as per standard behaviour) - and not throw an exception!
- there should be an option to opt-out
- and/or this should be extended to handle multiple generic arguments

**Comment by hisuwh** [comment]
@jbogard any chance of a 12.2.1 to fix https://github.com/jbogard/MediatR/pull/989

If and when I get some time I can look at the above as I have implemented generic handlers in a few projects.

**Comment by zachpainter77** [comment]
@jbogard  I have a fix for the above issue.. My only question is what limitations should be present?  What I mean is should we let the user define handlers that have 10 type parameters, each with no constraint? My solution is to allow more than one generic type parameter but throw errors if the max number of generic type params is exceeded OR the max number of concrete class combinations is exceed OR if a timeout occurs.  My question is.. what should those limitations be?

I'm also trying to figure out a way to test these scenarios since even defining a class that should throw an exception does so for all tests that call `AddMediatR`.  

So, any discussion around this is welcome.  I think I'm just going to create a PR and let discussion take place there.

**Comment by hisuwh** [comment]
Is there a way to define a sensible limit internally, but then allow for this to be configurable?

**Comment by zachpainter77** [comment]
yes, I think there is a way to do this.. I will set up some configuration options with default values that can be overridden when calling `AddMediatR`.. I will have a PR coming today. We can probably discuss what is sensible in terms of default limitations in that thread.

**Comment by jbogard** [comment]
This fix is on the MyGet feed: https://myget.org/gallery/mediatr-ci

Can you verify that this fixes the issue?

**Comment by zachpainter77** [comment]
![image](https://github.com/user-attachments/assets/7d0e1e5e-d5e9-4e52-91e7-2c4362e8654e)
 Seems to work with my test project. However I had to use this myget feed url in order to install the package via nuget package manager in visual studio.. 
 
 
![image](https://github.com/user-attachments/assets/569ba1a1-09b2-4dbd-a5f1-dc97f39835f4)

[https://www.myget.org/F/mediatr-ci/api/v3/index.json](https://www.myget.org/F/mediatr-ci/api/v3/index.json)

**Comment by jbogard** [comment]
Yes the pre release packages are all on MyGet. I'll get a release out then.

**Comment by jbogard** [comment]
Fixed by #1048
